### PR TITLE
[tdsql] The first draft of TDSQL-CDC

### PR DESCRIPTION
**Pull request by guixin**
TDSQL is an abbreviation for TecentDistributed SQL. **It is derived from MySQL** and designed like mysql cluster.

TDSQL CDC base on mysql cdc. It discovers tdsql's set and assigns them's splits to explicit readers for consumption. Fetcher Manager support multiple split reader thread when TM slot size lower then tdsql cluster set size.

TDSQL Document Link: https://cloud.tencent.com/document/product/557

### CODE REVIEW

**Comment by guixin** [comment]
@leonardBang Please take a look this pr. Thanks

**Comment by gong** [comment]
TDSQLhave TDSQL-C and TDSQL distribute mysql version. TDSQL distribute mysql version create table need shardkey. 
debezium will cause throw exception when using mysql-cdc-connector
```java
create table mytable( 
data_key int,
name varchar(255),
age int,
 PRIMARY KEY (`data_key`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8  shardkey=data_key': io.debezium.text.ParsingException: DDL 
mismatched input 'shardkey' expecting {<EOF>, '--'}
```
```java
ERROR io.debezium.pipeline.ErrorHandler [] - Producer failure
io.debezium.DebeziumException: Proxy ERROR: This type of SQL is not supported by group shard Error code: 602; SQLSTATE: HY000.
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1146) 
```
So, can this connector solve this problem?

**Comment by guixin** [comment]
> TDSQLhave TDSQL-C and TDSQL distribute mysql version. TDSQL distribute mysql version create table need shardkey. debezium will cause throw exception when using mysql-cdc-connector
> 
> ```java
> create table mytable( 
> data_key int,
> name varchar(255),
> age int,
>  PRIMARY KEY (`data_key`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8  shardkey=data_key': io.debezium.text.ParsingException: DDL 
> mismatched input 'shardkey' expecting {<EOF>, '--'}
> ```
> 
>  ```java
> ERROR io.debezium.pipeline.ErrorHandler [] - Producer failure
> io.debezium.DebeziumException: Proxy ERROR: This type of SQL is not supported by group shard Error code: 602; SQLSTATE: HY000.
> 	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1146) 
> ```
> 
> So, can this connector solve this problem?

Yes, tdsql cdc connector solve this problem. Welcome to try tdsql cdc connector. You need clone source to local, checkout feature/tdsql-cdc branch and compile it. If you encounter problems, please create issue to me, I will try solve it.

**Comment by guixin** [comment]
>

**Comment by guixin** [comment]
> > TDSQLhave TDSQL-C and TDSQL distribute mysql version. TDSQL distribute mysql version create table need shardkey. debezium will cause throw exception when using mysql-cdc-connector
> > ```java
> > create table mytable( 
> > data_key int,
> > name varchar(255),
> > age int,
> >  PRIMARY KEY (`data_key`)
> > ) ENGINE=InnoDB DEFAULT CHARSET=utf8  shardkey=data_key': io.debezium.text.ParsingException: DDL 
> > mismatched input 'shardkey' expecting {<EOF>, '--'}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```java
> > ERROR io.debezium.pipeline.ErrorHandler [] - Producer failure
> > io.debezium.DebeziumException: Proxy ERROR: This type of SQL is not supported by group shard Error code: 602; SQLSTATE: HY000.
> >  at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1146) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, can this connector solve this problem?
> 
> Yes, tdsql cdc connector solve this problem. Welcome to try tdsql cdc connector. You need clone source to local, checkout feature/tdsql-cdc branch and compile it. If you encounter problems, please create issue to me, I will try solve it.

tdsql cdc table ddl schema just like mysql cdc. the example is

```sql
Flink SQL> CREATE TABLE orders (
     order_id INT,
     order_date TIMESTAMP(0),
     customer_name STRING,
     price DECIMAL(10, 5),
     product_id INT,
     order_status BOOLEAN,
     PRIMARY KEY(order_id) NOT ENFORCED
     ) WITH (
     'connector' = 'tdsql-cdc',
     'hostname' = 'localhost',
     'port' = '3306',
     'username' = 'root',
     'password' = '123456',
     'database-name' = 'mydb',
     'table-name' = 'orders');
```

File: flink-sql-connector-tdsql-cdc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                                <relocation>
+                                    <pattern>com.google</pattern>
+                                    <shadedPattern>
+                                        com.ververica.cdc.connectors.shaded.com.google
+                                    </shadedPattern>
+                                </relocation>
+                                <relocation>
+                                    <pattern>io.grpc</pattern>
+                                    <shadedPattern>
+                                        com.ververica.cdc.connectors.shaded.io.grpc
```
Comment (gong) [inline flink-sql-connector-tdsql-cdc/pom.xml:74]: Hi, here include and relocation is not right. Maybe, you should refer to pom of flink-sql-connector-mysql-cdc.

Comment (guixin) [reply flink-sql-connector-tdsql-cdc/pom.xml]: @gong Thanks for point out this problem. I fixed it now. Did you find other problem? by the way, can you help me do code review? thanks a lot!

Comment (gong) [reply flink-sql-connector-tdsql-cdc/pom.xml]: > @gong Thanks for point out this problem. I fixed it now. Did you find other problem? by the way, can you help me do code review? thanks a lot!

@guixin I encounter network problems when I use this connector to test TDSQL districbute database. So I  can't conntiune to verify debezium exception problem. I think that debezium exception should need modify debezium to solve problem.

Comment (gong) [reply flink-sql-connector-tdsql-cdc/pom.xml]: > @gong network problems? Is flink runtime can't connect to tdsql set in cluster? use `/*proxy*/show status;` command to find tdsql cluster set info and try to connect one. example from TDSQL product wiki: https://cloud.tencent.com/document/product/557/8768
> 
> ```shell
> mysql> /*proxy*/show status;
>     +-----------------------------+------------------------------------------------------------------------------+
>     | status_name                 | value                                                                        |
>     +-----------------------------+------------------------------------------------------------------------------+
>     | cluster                     | group_1499858910_79548                                                       |
>     | set_1499859173_1:ip         | 10.49.118.165:5025;10.175.98.109:5025@1@IDC_4@0,10.231.23.241:5025@1@IDC_2@0 |
>     | set_1499859173_1:hash_range | 0---31                                                                       |
>     | set_1499911640_3:ip         | 10.49.118.165:5026;10.175.98.109:5026@1@IDC_4@0,10.231.23.241:5026@1@IDC_2@0 |
>     | set_1499911640_3:hash_range | 32---63                                                                      |
>     | set                         | set_1499859173_1,set_1499911640_3                                            |
> ```
> 
> try to connect set_1499859173_1: `mysql -h10.49.118.165 -P5025 -uxxx -pxxx`

I have network problem when I use public address to test.

**Comment by yuxiqian** [comment]
Hi @guixin, Thanks for your great work! Also sorry for the delay. Could you please rebase it to latest `master` branch since there’s been lots of changes in Flink CDC repo since your original commit? Kindly reminder that `com.ververica.cdc.connectors` package has been moved to `org.apache.flink.cdc.connectors`, and your new connector should be placed under `flink-cdc-connect/flink-cdc-source-connectors` path.

**Comment by guixin** [comment]
> Hi @guixin, Thanks for your great work! Also sorry for the delay. Could you please rebase it to latest `master` branch since there’s been lots of changes in Flink CDC repo since your original commit? Kindly reminder that `com.ververica.cdc.connectors` package has been moved to `org.apache.flink.cdc.connectors`, and your new connector should be placed under `flink-cdc-connect/flink-cdc-source-connectors` path.

Thanks for the reminder. I closed this pull request firstly

**Pull request closed without merging**